5주차 미션 제출

jambottle opened the pull request:
# 5주차 미션 제출 (김재원)

## Code Presentation

[영상 보기](https://vimeo.com/677428830)

## 구현 화면

### 상품 상세 정보 조회 (`/item/:itemNo`) 및 장바구니 목록 추가 결과
<img width="400" alt="week5_01" src="https://user-images.githubusercontent.com/72926450/153797974-b177f559-0718-4849-9754-e1f2c47287e4.png"> <img width="400" alt="week5_02" src="https://user-images.githubusercontent.com/72926450/153797986-36e0c059-aeaa-42dc-8021-9b146d92ee90.png">

### 장바구니 목록 조회 (`/cart`) 및 결과
- 똑같은 상품 3개가 장바구니 목록에 담겼을 때의 경우
- 주문 목록의 리스트 렌더링 및 총 결제 금액 계산 기능 구현

<img width="400" alt="week5_03" src="https://user-images.githubusercontent.com/72926450/153797988-5374d051-62c0-4128-8d6a-b7c686b1e514.png"> <img width="400" alt="week5_04" src="https://user-images.githubusercontent.com/72926450/153797992-138804dc-0da4-499c-8856-1fc216ceb001.png">

### 상품 주문 및 결제 (`/order` & `/order/complete`)
- 똑같은 상품 1개가 장바구니 목록에 담겼을 때의 경우
- 결제 정보 입력 시, 입력 값마다 유효성 검사 실시 기능 구현

<img width="400" alt="week5_05" src="https://user-images.githubusercontent.com/72926450/153797997-c6c306e5-daf0-4e4b-855c-1a53b4202988.png"> <img width="400" alt="week5_06" src="https://user-images.githubusercontent.com/72926450/153798000-4d8a9869-625b-45c5-b9c4-10f9f8fbd53b.png">

## 접근 방법

* [x] 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.

**📌 Vuex에 관한 요구사항을 정확히 숙지하고, 그에 따라 페이지와 컴포넌트를 개발하였습니다.**
- 상태 관리 라이브러리인 **Vuex를 적극 활용하고, 장바구니 API 통신 메서드와 그 결과로 받아온 장바구니 목록 정보 등을 전역 상태로 관리**
  - **`siteName`**: _(Vuex 활용을 연습하는 차원에서)_ AppHeader 컴포넌트의 문구를 전역 상태로 관리 및 수정
  - **`cartList`**: 기존에는 CartList 페이지에서 Method와 Lifecycle Hooks를 활용하여 관리하던 장바구니 API 통신 메서드와 그 결과를 Vuex Store로 옮기고, 역으로 CartList 페이지가 Vuex Store에서 전역 상태로 관리되는 메서드(`actions`)와 통신 결과(`state`)를 가져다 사용할 수 있도록 **Flux Pattern을 전방위적으로 도입하게 된 기준점 역할의 데이터**
    - **`getCartList`**: 비동기적인 API 통신 메서드로, **`cartList`** 초기 값을 장바구니 API로부터 수신하는 역할 (`actions`)
    - **`setCartList`**: 동기적인 메서드로, **`getCartList`** 통해 수신한 장바구니 목록을 **`cartList`** 내에 주입하는 역할 (`mutations`)
- 장바구니 목록 정보는 CartList 페이지뿐만 아니라 상품 주문 및 결제(OrderCheckout) 시에도 Vuex Store를 통해 활용 및 응용 가능
  - `mutations` 속성을 활용하면 새로운 상품 정보가 장바구니 목록에 추가될 수 있으며, `getters` 속성을 활용하면 장바구니에 담긴 상품들의 총 결제 금액을 계산 가능
    - **`postCartItem`**: 동기적인 메서드로, **`cartList`** 목록에 새로운 상품의 정보를 추가하는 역할 (`mutations`)
    - **`cartTotalPrice`**: computed와 거의 흡사한 속성으로, 장바구니에 담긴 상품들의 총 결제 금액을 계산하는 역할 (`getters`)
  - 이때, Vuex Store를 제외한 곳에서는 장바구니 관련 API를 직접 연동하는 경우가 없도록, 관련 API 및 상태를 전부 Vuex Store로 이전
  - Vuex Store의 편의성에 과도하게 의존할 경우의 문제점에 대해 인지하고, **불필요한 상태 관리에 관해서는 최대한 지양하는 방향을 선택!**

**📌 Vue Router에 관한 요구사항을 정확히 숙지하고, 이 기능에 대한 테스트 케이스를 포함한 T/C를 작성하였습니다.**
- 장바구니 목록(CartList) 페이지와 상품 주문 및 결제(OrderCheckout) 및 결제 완료(OrderComplete) 페이지마다 **다음 화면으로 넘어갈 수 있는 링크를 `<router-link>` 태그를 활용하여 구현하고, 이 기능을 포함한 내용들을 검증하는 테스트 코드를 파일별로 작성**
  - 각각의 페이지 및 하위 컴포넌트 파일들에 대해서는 라우팅과 리스트 렌더링 등에 대한 테스트 코드를 요구사항에 맞춰 작성
  - 앞서 작성한 테스트 코드를 기반으로 문서를 구성하되, 테스트에 실패할 경우에는 관련 내용을 정정하여 요구사항을 만족

**📌 최종적으로, "하나의 온전한 애플리케이션을 완성한다"는 목표 아래 다양하고 세세한 부분까지 작업 중에 있습니다.**
- **[스타일링]**
  - 각각의 라우팅 페이지별로 일관된 스타일을 추구하고 있으며, 특히 일부 Flexbox System을 Grid System으로 대체
  - 각 페이지별로 라우팅 될 때마다 하단 Footer Navigation Bar의 라우팅 경로와 일치할 때에는 버튼 활성화 기능 구현
- **[유효성 검사]** 결제 정보 입력 폼의 경우, 필수 입력과 같은 입력 조건을 설정할 필요가 있어 `vee-validate` 라이브러리로 유효성 검사 구현
  - _그 외에도 아직 구현하지 못한 **찜 목록, 마이 페이지**와 같은 영역에 대해서도 애플리케이션의 완성도를 위해 추가 기획을 병행 중_

## 특이사항

* [x] 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.

> 👍 하나의 완성된 애플리케이션의 모습을 갖춰나가는 게 눈으로 보이니까 보다 많은 기능을 구현해 완성도 높은 애플리케이션으로 만들고 싶은 욕심이 저절로 생겨나고 있습니다. **앞으로 더 구현하고자 하는 기능들을 리스트-업해서 하나씩 마저 구현해볼 수 있도록 해야겠습니다!** 🤠

> 👎 현시점에서 테스트를 실행할 때 너무 많은 에러 메시지가 발생해서 아직 Vuex에 대한 T/C를 작성하지 못했습니다...
에러 메시지를 해결한 뒤에 **[Testing Vuex](https://test-utils.vuejs.org/guide/advanced/vuex.html)를 참고하여 필요한 테스트 케이스를 충분히 채워 넣도록 하겠습니다!!**

𝑸. 4주차 미션과 이번 5주차 미션을 거치면서 Routing & Flux Store에 대해서까지 테스트를 할 수 있다는 사실과 그럼에도 반드시 테스트 기반으로 개발해야 하는 것이 맞다는 생각에 동시에 맞닥뜨리게 되니까... 여태까지 Unit Test 자체를 받아들이는 것에도 익숙지 못했는데, **이게 과연 실무에서는 어떤 수준까지 범위가 확장되어 적용되고 있는지가 궁금해졌습니다.**

저도 이와 관련된 자료나 아티클을 찾아보고 있는 중인데, 강사님께서도 **혹시 저희가 과정이 끝나고 난 뒤에도 Jest, Cypress와 같은 테스팅 프레임워크에 관해 더 깊이 공부할 수 있는 방법(또는 자료 등)을 추천해주실 수 있을까요?** 개인적으로는 이 과정에서 Jest를 배운 덕분에 앞으로도 해당 프레임워크를 적극 활용한 F/E 개발자가 되고 싶다는 생각이 들어서요...😅

𝑸. 저는 React로 먼저 F/E 개발을 접했고, React의 시장 내에서의 인기나 커뮤니티 규모 등의 장점을 인지하고 있으면서도 Vue.js에서의 개발 경험이 매우 편하고 좋다고 생각하고 있었습니다. 하나의 `.vue` 파일 내에서 문서 작성, 로직 구현, 스타일링을 깔끔하게 진행할 수 있다는 점에서 특히 그런 느낌을 많이 받았는데요. 그 와중에 **Composable API가 새롭게 출시되면서 기존에 Options API가 주던 규칙과 새로운 규칙이 서로 섞이는 것 때문에 개발 경험이 오히려 불편해진다는 느낌을 받았어요.**

F/E 개발자라면 새롭게 등장하는 신기술들에 빠르게 적응해야 한다는 것에 대해서도 알고 있지만, Options API를 기본으로 개발하는 상황에서 Composable API의 비중이 너무 높아지면 오히려 Vue.js에서 느껴왔던 장점이 상쇄되겠다 싶다는 생각이 드는데요... 특히, 이번 과정에서 동료 분들께서 Composable API를 쓰시는 경우를 많이 보게 되면서 뭐가 맞는 건지 더 헷갈리더라고요 🤔 **혹시 강사님께서는 Composable API에 대해서 어떤 견해를 가지고 계신지도 궁금하고, 실제로 실무에서는 적극적으로 도입되고 있는 추세인지도 매우 궁금합니다!**

(아직 Composable API를 써본 경험이 많지 않은 상태에서의 개인적인 직감으로는 **React에서의 `Custom Hook`** 같이 재사용성이 높은 로직에 한해 모듈화 하는 정도로만 사용하는 것이 적절해 보이긴 합니다...🧐)

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

jambottle committed:
Add: 라우팅 파일 모듈화 (3)
- 상품 주문 및 결제와 관련된 Order 메뉴에 대해서도 라우팅 파일을 분리하여 관리

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (1) CartList 페이지 및 Card 컴포넌트 레이아웃 구현
- 유사한 레이아웃 및 기능을 지닌 ItemList 페이지 및 Card 컴포넌트를 재사용하고, 연관된 스타일 수정

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (2) CartList 페이지 내에서 결제 화면으로 이동하는 버튼 추가

jambottle committed:
Modify: CardList 페이지 내에서 Card 컴포넌트의 너비 값 지정 방식 변경

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (3) 결제 완료 페이지 레이아웃 구현 완료 (OrderComplete.vue)

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (4) 결제 페이지 레이아웃 구현 완료 (OrderCheckout.vue)

jambottle committed:
Fix: OrderCheckout/OrderComplete 페이지 내에서 버튼 여백으로 발생한 오류 해결

jambottle committed:
Chore: vee-validate 패키지 설치

jambottle committed:
Add: OrderCheckout 페이지 내에서 입력되는 결제 정보에 대한 유효성 검사 1차 구현
- vee-validate 패키지를 활용하여, 우선 문자 입력에 대한 유효성부터 검증하고, 검증 실패 시 경고 메시지 송출

jambottle committed:
Modify: OrderCheckout 페이지의 전반적인 스타일 수정 완료 등

jambottle committed:
Test: Card 컴포넌트에 대한 단위 테스트 실시 (CartList)
- setProps 메서드로 주입한 임의의 데이터를 바탕으로 요구사항을 만족하는 기본적인 렌더링 여부를 확인
- 필요한 라우팅 룰을 불러와 상품의 상세 정보 페이지로 전환시키는 과정을 테스트 내에서 구현하여 검증

jambottle committed:
Test: CartList 페이지에 대한 단위 테스트 실시
- Jest에서 제공하는 fn 함수를 활용하여 API 통신 메서드를 Mocking한 뒤, 메서드의 호출 여부를 검증
- setData 메서드로 주입한 임의의 데이터를 바탕으로 요구사항을 만족하는 리스트 렌더링 여부를 확인

jambottle committed:
Test: 상품 주문 및 결제 페이지에 대한 단위 테스트 실시 (OrderCheckout.vue)
- 페이지 전환에 필요한 링크 및 버튼 요소의 존재 여부를 바탕으로 요구사항을 만족하는 렌더링 여부를 확인
- 필요한 라우팅 룰을 불러와 주문 및 결제 완료 페이지로 전환시키는 과정을 테스트 내에서 구현하여 검증

jambottle committed:
Test: 상품 주문 및 결제 완료 페이지에 대한 단위 테스트 실시 (OrderComplete.vue)
- 페이지 전환에 필요한 링크 및 버튼 요소의 존재 여부를 바탕으로 요구사항을 만족하는 렌더링 여부를 확인
- 필요한 라우팅 룰을 불러와 첫 페이지(상품 목록)로 전환시키는 과정을 테스트 내에서 구현하여 검증

jambottle committed:
Test: OrderCheckout/OrderComplete 페이지 내 data-test 속성 재정비
- 테스트 코드의 변동은 없으며, 요구사항에 맞춰 테스트 코드를 확정한 뒤 data-test 속성의 위치와 값을 조정

jambottle committed:
Modify: Header 컴포넌트에 대한 테스트 코드 내에서 테스트 케이스 설명 수정
- 기존 테스트 케이스에 작성해둔 설명이 테스트 코드의 내용을 충분히 담고 있지 못한 것을 발견하여 정정

jambottle committed:
Modify: ItemInfo 페이지에 대한 테스트 코드 내에서 테스트 케이스 내용 수정
- 해당 페이지 내의 버튼 문구를 UX 측면에서 변경하게 되면서 테스트에 통과할 수 있도록 관련 내용을 정정

jambottle committed:
Modify: Vuex Store 활용 연습 - (1) state로 Header 문구 수정
- 새롭게 설정한 전역 상태(siteName)로 Header 컴포넌트의 문구를 대체

jambottle committed:
Modify: Vuex Store 활용 연습 - (2) mapState 헬퍼 함수 적용

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (5) Vuex Store에 장바구니 목록 저장 기능 구현
- 장바구니 API의 활용을 Vuex Store로 옮기고, 장바구니 목록 정보를 state에 저장

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (6) state로 CartList 페이지의 장바구니 목록 정보 대체

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (7) state에 저장한 장바구니 목록 정보를 결제 화면에서 출력
- 결제 화면으로 넘어가더라도, Vuex Store에 저장되어 있는 장바구니 목록 정보를 리스트 렌더링으로 반복

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (8) 결제 화면에 출력된 장바구니 목록 정보에서 총 결제 금액을 계산
- state를 가공할 수 있게 하는 getters 속성으로 장바구니 목록의 총 결제 금액을 계산 후, 결제 화면에 출력

jambottle committed:
Chore: CartList 페이지 내 Card 컴포넌트의 paragraph 요소 삭제

jambottle committed:
Chore: CartList 페이지 내 Card 컴포넌트 및 button 요소 사이의 간격 조정

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (9) Vuex Store에 장바구니 목록 추가 기능 구현
- Vuex Store 내 mutations 속성으로 장바구니 목록에 새로운 상품의 정보를 추가하는 함수를 등록
- 해당 함수는 ItemInfo 페이지에 클릭 이벤트로 등록되어 상품의 정보를 장바구니 목록에 추가

jambottle committed:
Add: 5주차 미션 요구사항 구현 - (10) 장바구니 목록 정보를 저장하는 actions 이동
- CartList 페이지에서 장바구니 목록 정보를 불러오던 actions 속성의 getCartList 메서드의 오류 발견
- 새로고침 또는 페이지 전환으로 해당 페이지에 재접속하면, 장바구니 목록 정보가 초기화되는 문제 발생
- 해당 메서드의 위치를 최상위(App.vue)로 이동시켜 장바구니 목록이 모든 페이지에서 유지되도록 조치

jambottle committed:
Chore: CartList 페이지 내 Card 컴포넌트 및 button 요소 사이의 간격 삭제

jambottle committed:
Add: OrderCheckout 페이지 내에서 입력되는 결제 정보에 대한 유효성 검사 2차 구현
- 정규 표현식을 추가로 활용하여, 연락처 입력에 대한 유효성을 검증하고, 검증 실패 시 경고 메시지 송출

k12ms26 commented on mission/src/views/ItemInfo.vue:
```diff
@@ -40,17 +40,22 @@
+        "
         data-test="footer-button"
-        @click="showModal"
+        @click="showModal(), postCartItem(item)"
```
두 메서드 내의 기능을 하나의 메서드에 합쳐서 call하는 것이 더 보기 좋을 것 같은데, 혹시 나누신 이유가 있으실까요?? + 아니면 showModal 메서드 안 맨마지막 줄에 조건을 걸어서 postCartItem 메서드를 호출해도 괜찮을 것 같다고 생각했습니다!

jambottle replied:
안녕하세요, 민선님! 일단은 두 함수를 등록할 때 일부러 나눈 건 아니었어요 😃
실제로 '장바구니에 담기' 기능을 구현하게 될 줄 모르고, 예전에 Modal 컴포넌트와 함께 `showModal`이라는 함수를 구현해뒀는데요-

`showModal`이 Modal을 열어주는 역할의 지역 함수라면, `postCartItem`은 장바구니에 상품을 담아주는 역할의 전역 함수잖아요?
그래서 각기 다른 역할의 두 함수를 굳이 합치지 않고, 하나의 이벤트가 작동할 때에 두 개의 함수가 실행되게끔 구현한 거였어요 😯

jambottle replied:
제안해주신 내용에 대해서 조금 더 고민해보자면,

- 우선 `showModal`이라는 이름 아래에 `postCartItem`을 넣어주는 건 의미적인 차원에서 아닌 것 같고,
- 오히려 `showModal`이라는 이름을 바꾸고, 두 함수의 역할을 하나로 합쳐주는 방법은 적극 고려해볼 수 있을 것 같아요!!
_(제가 함수의 역할이 명확히 구분되고, 그 역할에 맞춰 이름 짓는 걸 되게 중요하게 생각해서요오...)_

jambottle replied:
제안해주신 내용을 적용해보려고 아래와 같이 시도를 해보았어요 😁

일단, 두 함수의 기능을 하나로 합치되, 함수 이름을 `showModal`에서 `postCartItemWithModal`로 바꿔서
`showModal`과 `postCartItem` 두 함수의 기능이 하나로 합쳐져 있다는 사실을 함수의 이름을 통해서 드러내려고 노력했어요!

```
postCartItemWithModal() {
  this.postCartItem(this.item);
  this.isModalShown = true;
},
```

그리고 아래와 같이 메서드 바인딩 방식을 바꿔줬는데요, 막상 민선님 말씀처럼 수정해보니까 훨씬 깔끔하고 좋네요 👏👏👏

```suggestion
        @click="postCartItemWithModal"
```

k12ms26 replied:
@jambottle 이렇게 자세하게 써주시다니 👍 👍 두 메서드의 역할이 확실하게 다른건 있네요!!
제 의견 반영해서 수정해주신 것 넘넘 감사합니다 😄 저도 배웠어요

jambottle replied:
아뇨, 민선님이 질문 주신 덕분에 한 번 더 생각해볼 계기가 생겨서 좋았습니다 🙏

k12ms26 reviewed (commented):
[좋은 점/잘한 점]
저는 해당 상품이 장바구니에 담겼으면 아예 장바구니에 담지 못하도록 조건을 걸어주었는데, 생각해보니 수량 등을 생각하면 기능이 더 확장되었을 경우에는 그렇게 막아두면 안되겠네요 😮 ... 재원님 pr 내용 보고 깨달았습니다 감사해요

[의견제시]
코멘트 달아두었습니다 :)

daaeaad commented on mission/src/store/index.js:
```diff
@@ -1,8 +1,49 @@
+        sum += state.cartList[i].price;
+      }
+      return sum;
+    },
```
반복문은 forEach 메서드 사용을 고려해보시는 것도 좋을 것 같아요 :)
선호도 차이라고 하는데 개인적으로 중간에 break가 필요한 조건이 없다면 좀더 간결해 보이더라구요!

jambottle replied:
다애님, 오랜만이네요! 지난 5주 동안 다애님도 정말 수고 많으셨어요 :-D

여기 제안해주신 내용은 바로 적용해볼 수 있을 것 같아요!! 아래와 같이 `forEach`로 구현하는 걸 말씀하신 거죠?
아무래도 제가 쉽고 익숙한 메서드를 써 버릇하는 경향이 있어서 별 고민 없이 `for`을 사용했던 것 같네요 😆

코드 라인도 한 줄로 줄어들어 확실히 코드가 훨씬 간결해지고 좋은 것 같습니다 👍👍

```suggestion
    cartTotalPrice(state) {
      let sum = 0;
      state.cartList.forEach((item) => { sum += item.price; });
      return sum;
    },
```

daaeaad reviewed (commented):
- notion에 그동안의 작업사항을 꼼꼼히 정리하신 점이 돋보였어요!
- 테스트 코드도 꼼꼼히 작성하신 것 같습니다 👍  추후 vuex 테스트 코드도 잘 작업하실 수 있을 것 같아요 :)
- 폼에 입력받는 데이터 유효성 검사를 진행하심으로써 작업물의 완성도가 더 올라간 것 같아요!

**[ 의견 ]**
- 코멘트 남겨두었습니다. 5주 동안 고생 많으셨어요 :)

jambottle committed:
Refactor: CartList 페이지 내 Card 컴포넌트의 paragraph 요소 삭제로 불필요해진 코드 제거

jambottle committed:
Refactor: 피어 리뷰 반영 - (1) 메서드 및 바인딩 관련 코드 간소화
- 메서드가 두 개씩 바인딩된 경우, 두 메서드를 하나의 새로운 메서드로 합치는 작업 수행 (ItemInfo.vue)
- Vuex Store 내의 getters 중에서 for를 활용한 반복문을 forEach를 통해 간결하게 만드는 작업 수행

jambottle commented on mission/src/views/CartList.vue:
```diff
@@ -1,64 +1,72 @@
+    CartListCard,
   },
 
-  methods: {
-    async getCartList() {
-      try {
-        const { data, status } = await CartRepository.getCartList();
-        if (status !== 200) {
-          throw new Error('장바구니의 목록을 조회할 수 없습니다.');
-        } else {
-          this.items = data.cart_item;
-        }
-      } catch (err) {
-        console.error(err);
-      }
-    },
-  },
-
-  created() {
-    this.getCartList();
```
해당 부분이 전역 상태 관리 영역에 포함되면서 `getCartList`는 **Vuex Store로 옮겼고,**
`created() { this.getCartList() }`부분 역시 **최상위 컴포넌트인 `App.vue`로 이동시켰습니다.**

jambottle replied:
해당 부분을 전역 상태 관리의 영역으로 포함하게 되면서 **`getCartList`는 Vuex Store로 옮겼고,**
`created() { this.getCartList() }` 부분 역시 **최상위 컴포넌트인 `App.vue`로 이동시켰습니다.**

Subinhyun commented on mission/src/components/CartList/Card.vue:
```diff
@@ -0,0 +1,108 @@
+    discountPrice() {
+      return this.discount_price.toLocaleString();
+    },
+    originalPrice() {
```
사용하지 않는 함수면 지워도 될 것 같습니다!

jambottle replied:
아, 그렇네요! 비슷한 컴포넌트를 복사해서 쓰다 보니까 몰랐어요... 꼼꼼히 봐주셔서 감사합니다 🙌

Subinhyun reviewed (commented):
기능 구현을 잘 해주셨는데 PR 내용도 자세해서 이해하기 쉬웠습니다! 
꼼꼼하게 정리도 잘 해주셔서 배울 점이 정말 많네요 ㅜ
바로바로 피드백을 반영하는 점도 대단하셔요 ㅠㅠ 👍🏻 수고하셨습니다!

allegoryj commented:
재원님 안녕하세요! 이번 미션도 고생많으셨습니다. 요구사항 뿐만 아니라 추가 기능들까지 구현을 잘 구현해주시고 vuex 관련 개념도 영상에서 잘 설명해주셔서 도움이 많이 되었습니다. commit 메세지도 항상 명료하게 남겨주셔서 개발 과정을 살펴보기 좋았습니다. 저는 이번 미션 TC 작성을 못했는데 테스트 코드도 견고하게 작성하신 것 꼭 참고해보도록 하겠습니다. 재원님이 그동안 남겨주신 피드백들이 개인적으로 도움이 참 많이되어 감사한 마음입니다. 5주간 정말 고생많으셨습니다! :)

ohyein00 commented on mission/tests/unit/Order/OrderCheckout.spec.js:
```diff
@@ -0,0 +1,51 @@
+      },
+    });
+
+    await container.get('[data-test="order-router"]').trigger('click');
+    await flushPromises();
```
flushPromises라는 것을 처음 보게 되어서 간략히 찾아보니 비동기 처리를 위함이라고 이해했습니다!
그런데 .trigger('click) 해당 단계에 이미 await으로 비동기처리가 되어있음에도 flushPromises를 쓰신 이유가 궁금합니다.어떤 차이가 있는건가요??😮재원님 덕에 새로운 것을 알아갑니다. 감사합니다!

jambottle replied:
음... 이거 아마도 클릭 이벤트로 인해 다른 페이지로 라우팅 되는 경우에 써줬던 걸로 기억하는데요 🤔

await이 붙어 있기는 하지만, 라우팅이 되면서 API Call과 같은 비동기 이벤트가 바로 실행되는 경우를 대비해서
_'뭐가 추가로 실행될 예정이니까 조금만 더 기다렸다가 테스트 결과를 확인해'_ 같은 이유로 저 메서드를 활용한다고 알고 있어요-

저도 어떤 예제에서 보고 적용해본 메서드라서 없어도 동작하는지에 대해서는 다시 체크해봐야 할 것 같아요!!

ohyein00 commented:
- pr의 내용을 정말 꼼꼼하게 잘 작성해주시고, vuex 상태관리 방법에 대해서도 자세히 남겨주셔서 많은 참고가 되었습니다! 정말 열심히 공부하신 것 같아요 대단합니다 👍
- 저는 그냥 컴포넌트에서 장바구니 계산 메소드를 만들었는데, 재범님께서 하신 것처럼 getter을 이용하는 것이 재사용측면에서 더 좋을 것이라고 생각이 드네요! 
- 커밋 메세지를 잘 남겨주셔서 코드를 이해하기 편했습니다. 재범님의 커밋 습관을 저도 배워가야겠습니다

glowing713 commented on mission/tests/unit/CartList/CartList.spec.js:
```diff
@@ -0,0 +1,70 @@
+    expect(CartRepository.getCartList).toHaveBeenCalledTimes(1);
+  });
+
+  it('renders multiple CartListCard components', async () => {
```
mock 데이터를 정의하셔서 각 상품이 잘 렌더링이 되는지 테스트를 하셨군요..!! 이 부분도 정말 중요한 부분인데 잊었던 것 같습니다! 더욱 견고하게 테스트코드를 작성하시는 부분 배워갑니다..!!!👍🏻👍🏻

jambottle replied:
기본적인 렌더링 부분에 대해서는 1주차 때부터 익숙해진 게 있다 보니까 잘할 수 있겠는데,
비동기 & 라우팅 & Vuex... 이런 단계의 테스트 코드를 작성하는 데에는 아직 한참 부족합니다 🥲

그래도 좋은 말씀 해주셔서 힘이 나네요! 선환님도 저도 Jest와의 인연을 놓지 말아 보아요 ㅎㅎㅎ

glowing713 reviewed (commented):
[잘한 점]
- PR 내용과 Notion에 정리하신 내용들을 보면서 재원님의 열정이 느껴지는 것 같습니다..!! 특히 PR 내용만 보아도 이번 주차에 어떤 기능을 개발하셨는지 한눈에 알아볼 수 있을 것 같아요!
- 유효성검사 부분도 많이 배워갑니다😃 감사드려요!

[의견제시]
- 이번 리뷰때 처음 뵙게 되었지만 재원님 코드를 보면서 정말 많은 내용을 배워갈 수 있었던 것 같습니다:) 5주동안 정말 고생 많으셨습니다!

2taesung commented:
노션에 정리 해놓으신거 보고 감탄했네요 .. 
저도 마음은 재원님처럼 정리하며 공부하려 했는데 반성하게 됩니다 ㅠㅠ
영상에서도 코드 작성에 대한 이유를 잘 설명해주셔서 이해가 잘됐고 큰 도움이 됐습니다.
vuex 뿐만 아닌 코드의 모듈화에 대해 많은 고민을 하신 것 같습니다.
5주동안 수고하셨습니다!

jambottle committed:
Refactor: 피어 리뷰 반영 - (2) 사용하지 않아 불필요해진 코드 제거 (CartList/Card.vue)

jambottle committed:
Refactor: 피어 리뷰 반영 - (3) Lazy Loading 적용
- 사이트 접속 시, 처음부터 로딩될 필요가 없다고 판단되는 Order/User/Wish 페이지에 Lazy Loading 적용 등

jambottle committed:
Refactor: 피어 리뷰 반영 - (4) Vuex Store 모듈화 (CartModule.js)

jambottle committed:
Refactor: 피어 리뷰 반영 - (5) Vuex Store 모듈화에 맞춰 헬퍼 함수의 활용 방식 변경
- Vuex Store 모듈화에서 namespaced 속성을 사용하게 되면서 createNamespacedHelpers 메서드를 추가 적용

lkaybob reviewed (commented):
@jambottle 안녕하세요, Code presentation 영상을 확인하고 리뷰 드리려고 하는데 열리지 않아서 리뷰 홀딩 중입니다. 자세한 건 디스코드에서 매니저님이 DM을 드렸다고 하니 확인 부탁드립니다.

jambottle committed:
Refactor: 피어 리뷰 반영 - (6) Vuex Store에 장바구니 목록 초기화 기능 구현
- Vuex Store 내 mutations 속성으로 장바구니 목록을 빈 배열로 초기화시키는 기능의 함수를 등록
- 해당 함수는 OrderCheckout 페이지에 클릭 이벤트로 등록되어 결제가 완료되면 장바구니 목록을 초기화

jambottle committed:
Test: CartList/OrderComplete 페이지 내 data-test 속성 재정비

jambottle committed:
Refactor: 피어 리뷰 반영 - (7) 장바구니 목록 초기화 시, 장바구니 목록을 대체할 내용 구현
- Vuex Store 내 state 속성으로부터 불러온 장바구니 목록의 상품 유무를 기준으로 조건부 렌더링 구현
- 만약 장바구니 목록이 초기화되어 배열의 요소가 없을 경우, 안내 메시지 및 '쇼핑하러 가기' 버튼 노출

jambottle committed:
Refactor: 피어 리뷰 반영 - (8) 결제 정보 입력 불충분 시, '결제하기' 버튼 비활성화 기능 구현
- 입력되는 결제 정보를 양방향 바인딩으로 관리하고, computed 속성으로 입력 상황을 실시간으로 검사
- 만약 결제 정보 입력이 충분하지 않을 경우, disabled 속성을 활용해 '결제하기' 버튼을 비활성화 처리

jambottle committed:
Test: OrderCheckout/OrderComplete 페이지에 대한 테스트 코드 내 data-test 속성 재정비

lkaybob reviewed (commented):
(코드리뷰 코멘트가 누락이 되었네요. 늦게나마 제출일(13일) 기준 커밋들 기준으로 코멘트드립니다. 늦어서 죄송합니다.)

구현을 늦게 시작하신 것 같아 제출일에는 외관 완성을 위주로 진행을 하신 것 같습니다. 이후 히스토리로 추정했을 때는 store를 도입하여 마저 구현을 해주신 것 같네요. 추가적인 기능들을 구현하셔서 더 보강하시면 좋은 결과물이 나오지 않을까 합니다. 5주동안 고생하셨습니다.
